firestore/client/field_path.h

```diff
@@ -0,0 +1,73 @@
+// Copyright 2018 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLECLOUDCPP_FIRESTORE_GOOGLE_FIRESTORE_FIELD_PATH_H_
+#define GOOGLECLOUDCPP_FIRESTORE_GOOGLE_FIRESTORE_FIELD_PATH_H_
+
+#include <iostream>
+#include <regex>
+#include <string>
+#include <vector>
+
+namespace firestore {
+/**
+ * A FieldPath refers to a field in a document. The path may consist of
+ * a single field name (referring to a top level field in the document),
+ * or a list of field names (referring to a nested field in the document).
+ */
+class FieldPath {
+ public:
+  // Construct FieldPath from a vector of field names
+  FieldPath(const std::vector<std::string>& parts);
+
+  // Construct FieldPath from a field path string
+  static const FieldPath from_string(const std::string& string);
+
+  // The regex for a simple field name
+  static std::regex simple_field_name;
+
+  // Construct a new FieldPath by appending a field path string
+  const FieldPath append(const std::string& path) const;
```

Comment:
Returning a `const` has no effect, they can be assigned to non-const objects.  Please remove here and elsewhere.

Reply:
@coryan I have addressed your other comments.  I recalled that there was a difference returning const objects, which I have found through [here](https://www.linuxtopia.org/online_books/programming_books/thinking_in_c++/Chapter08_014.html).

Perhaps things have changed or that it is irrelevant in this case?

Reply:
Thanks!  I learned something today.  I remain ~stubborn~ unconvinced.  Your type is closer to an `int`, it is a value type with only `const` member functions.  What difference does it make to the user to return a `const FieldPath`?  And why not simplify the code if it makes no difference?

Reply:
To try to convince you, what about `const` member functions but doesn't modify anything in the class?  Would you make the member function `const` or would you say it doesn't matter because it's not modifying the class?

for example, 
`size_t size() {return 0;}`
or 
`size_t size() const {return 0;}`?

I tend to think C++ is in lockdown mode and things should be const unless they have to not be.  
I did make the change in a commit though.